CRM/Activity/Tokens.php

```diff
@@ -98,20 +138,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
     }
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function prefetch(\Civi\Token\Event\TokenValueEvent $e) {
+    if (!empty($e->getTokenProcessor()->context['actionMapping'])) {
+      // Scheduled reminders provide results objects so don't prefetch
+      return NULL;
+    }
+
+    // Get the activities
+    foreach ($e->getRows() as $row) {
+      $activityIds[] = $row->context['activityId'];
+    }
+
+    $activities = civicrm_api3('Activity', 'get', array(
+      'id' => array('IN' => $activityIds),
```

Comment:
Might fail if processing a batch of 25+ messages. Probably needs `'options' => ['limit' => 0]`?

Reply:
Also -- the loop above (populating `$activityIds`) makes perfect sense, but let me put it under a little extra scrutiny because this patch may become an example to inform similar patches to `CRM/*/Tokens.php`. It'll probably be fairly common to extract a bunch of `activityId` values (or `contributionId` or `caseId` or whatever), and that multiplies out to several 5 SLOC bits of boilerplate. Instead, I'd vote to skip the variable `$activityIds` and have a helper function like:

```php
    $activities = civicrm_api3('Activity', 'get', array(
      'id' => array('IN' => $e->getTokenProcessor()->getContextValues('activityId')),
      'options' => ['limit' => 0]
    );
```

https://github.com/civicrm/civicrm-core/compare/master...totten:master-act-tok?expand=1#diff-a2c7ae339db570d160a69bd288516d11R171

(Aside: I've tried a few variations of `CRM_Utils_Array::collect()` and `array_map()`, but the magic bits made that more complicated. Adding `getContextValues()` seemed easier.)

Reply:
Thanks - I've added `getContextValues()`.  I still have `$activityIds` since it is used later in `ActivityContact.get`